<a href="https://colab.research.google.com/github/Soumee06/Deep-Learning/blob/main/ANN_Udemy_artificial_neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Artificial Neural Network

### Importing the libraries

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf

In [ ]:
tf.__version__

'2.9.2'

## Part 1 - Data Preprocessing

### Importing the dataset

In [ ]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values #we are excluding the first three columns as they have no impact
# so for x we keep all columns :, and start keeping all columns from 3rd index and we stop before the last column -1.
y = dataset.iloc[:, -1].values #here we keep only the last column

In [ ]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [ ]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
X[:, 2] = le.fit_transform(X[:, 2])

In [ ]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding(Dummy variable generation) the "Geography" column

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(ct.fit_transform(X))

In [ ]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train) #applying to all features of training set
X_test = sc.transform(X_test) #apllying to all features of test set

## Part 2 - Building the ANN

### Initializing the ANN

In [ ]:
ann = tf.keras.models.Sequential() #buiding an instance ann which is a sequence of layers
#keras library was integrated into tf 

### Adding the input layer and the first hidden layer

In [ ]:
# Dense is used to add a fully connected layer into our ann class
ann.add(tf.keras.layers.Dense(units=6,activation='relu' ))                                      #add is the method of sequential class
# layers is a module of keras library and from layers module we pick the Dense class
#n_units- no of neurons we want to have in our hidden layers
#rectifier i.e. relu is the activation function

### Adding the second hidden layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6,activation='relu' ))

### Adding the output layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) #no of neurons =1 as the output as one value
#the activation function is sigmoid 

## Part 3 - Training the ANN

# Compiling the model

In [ ]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy']) 
#the optimizer updates the weight , for binary predsiction the loss function is binary_crossentropy and
# the metrics for evaluation is entropy

For non binary optimization the loss function changes and even activation function for non binary will be soft max

### Training the ANN on the Training set

In [ ]:
ann.fit(X_train,y_train,batch_size=32,epochs=100)
#.fit method on ann object to train our model
#batch training is more efficient in artificial training since training in batches is efficient
#defualt value of batch training =32
#epoch= 100

Epoch 1/100
250/250 [==============================] - 1s 2ms/step - loss: 0.6303 - accuracy: 0.7251
Epoch 2/100
250/250 [==============================] - 0s 2ms/step - loss: 0.5043 - accuracy: 0.7960
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4505 - accuracy: 0.7964
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4321 - accuracy: 0.8071
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4204 - accuracy: 0.8111
Epoch 6/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4103 - accuracy: 0.8190
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4008 - accuracy: 0.8253
Epoch 8/100
250/250 [==============================] - 0s 2ms/step - loss: 0.3926 - accuracy: 0.8270
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.3866 - accuracy: 0.8270
Epoch 10/100
250/250 [==============================] - 0s 2ms/step - loss: 0.3819 - accura

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

**Homework**

Use our ANN model to predict if the customer with the following informations will leave the bank: 

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer?

In [ ]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))
# the predict method has to be called on the same scaled object "sc" and .transform method is called for test data
#France is 1,0,0 is dummy encoded

1/1 [==============================] - 0s 98ms/step
[[0.01970473]]


In [ ]:
# Since there is small probability of this customer's leaving the bank
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]]))>0.5)

1/1 [==============================] - 0s 16ms/step
[[False]]


we are choosing 0.5 as a threshold , idf the probability  is greater than 0.5 then we consider the customer will leave the bank i.e. 1, if the probability is less than 0.5 then customer will not leave the bank .


**Solution**

Therefore, our ANN model predicts that this customer stays in the bank!

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [ ]:
y_pred = ann.predict(X_test) #finding out predicted probabilities
y_pred = (y_pred > 0.5) #converting the predicted probabilities into boolean values 0 and 1
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

63/63 [==============================] - 0s 1ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


We get the actual and predicted test results side by side

### Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1530   65]
 [ 205  200]]


0.865

 We have 1526 correct predictions that the customer stays
 200 correct predictions that the customer leaves
 (65+205) incorrect predictions
 our accuracy score for correct predictions is 86.5%